internal/controllers/buildkit/reconciler.go

```diff
@@ -0,0 +1,200 @@
+//+kubebuilder:rbac:groups=buildkit.seatgeek.io,resources=buildkits/finalizers,verbs=update
+//+kubebuilder:rbac:resources=pods,verbs=get;list;watch;create;update;patch;delete
+
+const controllerName = "Buildkit"
+
+type state = types.State[*v1alpha1.Buildkit]
+
+type reconciler struct {
+	c      *io.ClientApplicator
+	scheme *runtime.Scheme
+	log    *zap.SugaredLogger
+}
+
+func (r *reconciler) runBuildkit() *state {
+	return &state{
+		Name:      "run-buildkit",
+		Condition: conditionDeployed,
+		Transition: func(ctx context.Context, obj *v1alpha1.Buildkit, out *types.OutputSet) (*state, types.Result) {
+			log := r.log.With("name", obj.Name, "namespace", obj.Namespace)
+
+			// Check if we already have any Buildkit pods
+			managedPods, err := r.getExistingManagedPods(ctx, obj, log)
+			if err != nil {
+				return nil, types.ErrorResult(err)
+			}
+
+			// Ensure we have exactly one Buildkit pod running, creating or deleting as necessary
+			pod, err := r.ensureExactlyOnePod(ctx, obj, managedPods, out, log)
+			if err != nil {
+				return nil, types.ErrorResult(err)
+			}
+
+			// Do we need to add or remove anything? If so, apply those changes now and requeue.
+			if out.GetApplied().Len() > 0 || out.GetDeleted().Len() > 0 {
+				return nil, types.Result{
+					Done:                   true,
+					RequeueAfterCompletion: true,
+					RequeueMsg:             "Applying pod changes",
+					Reason:                 "ApplyingChanges",
+				}
+			}
+
+			// At this point, we have exactly one pod, so we can proceed to check its status.
+			// We'll empty the endpoint field and only set it once we confirm the pod is running and healthy.
+			obj.Status.Endpoint = ""
+
+			// Are all containers running and healthy?
+			if pod.Status.Phase == corev1.PodFailed {
+				log.Warnw("Buildkit pod has failed", "pod", pod.Name, "reason", pod.Status.Reason, "message", pod.Status.Message)
+				return nil, types.Result{
+					Done: true,
+					CustomStatusCondition: &types.ResultStatusCondition{
+						Reason:  api.ReasonUnavailable,
+						Status:  corev1.ConditionFalse,
+						Message: fmt.Sprintf("Buildkit pod %s has failed: %s", pod.Name, cmp.Or(pod.Status.Message, pod.Status.Reason, "unknown failure")),
+					},
+				}
+			}
+
+			if pod.Status.Phase != corev1.PodRunning {
+				log.Debugw("Buildkit pod is not yet running", "pod", pod.Name, "phase", pod.Status.Phase)
+				return nil, types.RequeueResultWithReasonAndBackoff("Buildkit pod not running", "PodNotRunning")
+			}
+
+			for _, containerStatus := range pod.Status.ContainerStatuses {
+				if !containerStatus.Ready {
+					log.Debugw("Buildkit pod container not ready", "pod", pod.Name, "container", containerStatus.Name)
+					return nil, types.RequeueResultWithReasonAndBackoff("Buildkit pod container not ready", "ContainerNotReady")
+				}
+			}
+
+			// If we reach here, the pod is running and all containers are ready!
+			if len(pod.Spec.Containers) == 0 || len(pod.Spec.Containers[0].Ports) == 0 {
```

Comment:
Could we catch this issue in the validating webhook?

Reply:
We do have logic in the `BuildkitTemplate` default webhook to ensure at least one container is set during admission.

However, this code here is checking the `Pod` we loaded from `buildkitObj.Status.ResourceRefs`. While that _should_ always be one that was created from the template via our builder, it's _possible_ (but highly unlikely) somebody could modify those ref to point to some other malformed pod, which is why there's an extra check here to prevent an out-of-bounds panic.